Common/Thread/ParallelLoop.cpp

```diff
@@ -0,0 +1,142 @@
+#include <cstring>
+
+#include "Common/Thread/ParallelLoop.h"
+#include "Common/CPUDetect.h"
+
+class LoopRangeTask : public Task {
+public:
+	LoopRangeTask(WaitableCounter *counter, const std::function<void(int, int)> &loop, int lower, int upper)
+		: counter_(counter), loop_(loop), lower_(lower), upper_(upper) {}
+
+	void Run() override {
+		loop_(lower_, upper_);
+		counter_->Count();
+	}
+
+	std::function<void(int, int)> loop_;
+	WaitableCounter *counter_;
+
+	int lower_;
+	int upper_;
+};
+
+WaitableCounter *ParallelRangeLoopWaitable(ThreadManager *threadMan, const std::function<void(int, int)> &loop, int lower, int upper, int minSize) {
+	if (minSize == -1) {
+		minSize = 1;
+	}
+
+	int numTasks = threadMan->GetNumLooperThreads();
+
+	int range = upper - lower;
+	if (range <= 0) {
+		// Bad range. A finished counter allocated.
+		return new WaitableCounter(0);
+	}
+
+	if (range <= numTasks) {
+		// Just assign one task per thread, as many as we have.
+		WaitableCounter *waitableCounter = new WaitableCounter(range);
+		for (int i = 0; i < range; i++) {
+			threadMan->EnqueueTaskOnThread(i, new LoopRangeTask(waitableCounter, loop, i, i + 1), TaskType::CPU_COMPUTE);
+		}
+		return waitableCounter;
+	} else {
+		// Split the range between threads. Allow for some fractional bits.
+		const int fractionalBits = 8;
+
+		int64_t totalFrac = (int64_t)range << fractionalBits;
+		int64_t delta = totalFrac / numTasks;
+
+		delta = std::max(delta, (int64_t)minSize << fractionalBits);
+
+		// Now we can compute the actual number of tasks.
+		// Remember that stragglers are done on the current thread
+		// so we don't round up.
+		numTasks = (int)(totalFrac / delta);
+		printf("numTasks: %d\n", numTasks);
+
+		WaitableCounter *waitableCounter = new WaitableCounter(numTasks);
+		int64_t counter = (int64_t)lower << fractionalBits;
+
+		// Split up tasks as equitable as possible.
+		for (int i = 0; i < numTasks; i++) {
+			int start = (int)(counter >> fractionalBits);
+			int end = (int)((counter + delta) >> fractionalBits);
+			if (end > upper) {
+				// Let's do the stragglers on the current thread.
+				break;
+			}
+			threadMan->EnqueueTaskOnThread(i, new LoopRangeTask(waitableCounter, loop, start, end), TaskType::CPU_COMPUTE);
+			counter += delta;
+			if ((counter >> fractionalBits) > upper) {
```

Comment:
`>=`?  If counter hits upper (exclusive), we don't want to run another pass.  I suppose in that case `end > upper` would be guaranteed.

-[Unknown]

Reply:
Yeah, that's right, thanks.